src/autoprops.rs

```diff
@@ -0,0 +1,261 @@
+use crate::function_component::FunctionComponentName;
+use quote::{quote, quote_spanned};
+use syn::spanned::Spanned;
+
+#[derive(Clone)]
+pub struct Autoprops {
+    item_fn: syn::ItemFn,
+    properties_name: syn::Ident,
+}
+
+impl syn::parse::Parse for Autoprops {
+    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+        let parsed: syn::Item = input.parse()?;
+
+        let item_fn = match parsed {
+            syn::Item::Fn(m) => m,
+            item => {
+                return Err(syn::Error::new_spanned(
+                    item,
+                    "`autoprops` attribute can only be applied to functions",
+                ))
+            }
+        };
+
+        let syn::ItemFn { attrs, sig, .. } = &item_fn;
+
+        let mut component_name = item_fn.sig.ident.clone();
+
+        attrs
+            .iter()
+            .find(|attr| {
+                match &attr.meta {
+                    syn::Meta::Path(path) => {
+                        if let Some(last_segment) = path.segments.last() {
+                            if last_segment.ident == "function_component" {
+                                return true;
+                            }
+                        }
+                    }
+                    syn::Meta::List(syn::MetaList { path, tokens, .. }) => {
+                        if let Some(last_segment) = path.segments.last() {
+                            if last_segment.ident == "function_component" {
+                                if let Ok(attr) =
+                                    syn::parse2::<FunctionComponentName>(tokens.clone())
+                                {
+                                    if let Some(name) = attr.component_name {
+                                        component_name = name;
+                                    }
+                                }
+                                return true;
+                            }
+                        }
+                    }
+                    _ => {}
+                }
+                false
+            })
+            .ok_or_else(|| {
+                syn::Error::new_spanned(
+                    sig,
+                    "could not find #[function_component] attribute in function declaration \
+                     (#[autoprops] must be placed *before* #[function_component])",
+                )
+            })?;
+
+        for input in &sig.inputs {
+            if let syn::FnArg::Typed(syn::PatType { pat, .. }) = input {
+                if let syn::Pat::Wild(wild) = pat.as_ref() {
+                    return Err(syn::Error::new_spanned(
+                        wild,
+                        "cannot use `_` as field name",
+                    ));
+                }
+            }
+        }
+
+        let properties_name = syn::Ident::new(
+            &format!("{}Props", component_name),
+            proc_macro2::Span::call_site(),
+        );
```

Comment:
The span should be `component_name.span()`, not `call_site`. Also consider using [`format_ident`](https://docs.rs/quote/latest/quote/macro.format_ident.html)

Reply:
ah cool I didn't know. Done in 248b0c2c3c54889a690688c7cbb1a48ebf21f5d4